### ISSUE TITLE
[1.2.x] CrossVersion.Disabled doesn't work

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#4977]
## migration note

`CrossVersion.Disabled` doesn't work as expected in sbt 1.0.x ~ 1.2.x.

We recommend plugin authors to use `CrossVersion.disabled` instead. For pattern matching, you can either check equality with `CrossVersion.disabled` or type match with `case _: sbt.librarymanagement.Disabled`.

## steps 1

```scala
CrossVersion(CrossVersion.Disabled, "1.0.0", "1.0")
```

## problem 1

It doesn't compile.

```scala
<console>:15: error: type mismatch;
 found   : sbt.librarymanagement.Disabled.type
 required: sbt.librarymanagement.CrossVersion
       CrossVersion(CrossVersion.Disabled, "1.0.0", "1.0")
```

## steps 2

```scala
scala> val currentBinaryVersion = "1.0"
currentBinaryVersion: String = 1.0

scala>   def scalaJSMapped(cross: CrossVersion): CrossVersion = cross match {
           case CrossVersion.Disabled =>
             CrossVersion.constant("sjs" + currentBinaryVersion)
           case cross: sbt.librarymanagement.Constant =>
             cross.withValue("sjs" + currentBinaryVersion + "_" + cross.value)
           case cross: CrossVersion.Binary =>
             cross.withPrefix("sjs" + currentBinaryVersion + "_" + cross.prefix)
           case cross: CrossVersion.Full =>
             cross.withPrefix("sjs" + currentBinaryVersion + "_" + cross.prefix)
         }
scalaJSMapped: (cross: sbt.CrossVersion)sbt.CrossVersion

scala> scalaJSMapped(CrossVersion.disabled)
```

From https://github.com/scala-js/scala-js/blob/7c502f61c87735898c2293cb937eee8f55ad8686/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSCrossVersion.scala

## problem 2

```scala
scala> scalaJSMapped(CrossVersion.disabled)
scala.MatchError: Disabled() (of class sbt.librarymanagement.Disabled)
  at .scalaJSMapped(<console>:15)
  ... 36 elided
```

## expectation

```scala
scala> CrossVersion(CrossVersion.Disabled, "1.0.0", "1.0")
res3: Option[String => String] = None

scala> scalaJSMapped(CrossVersion.disabled)
res5: sbt.CrossVersion = Constant(sjs1.0)
```

## notes

This will be fixed in 1.3.0, but it doesn't currently work in 1.2.x.

- sbt/librarymanagement#121 added `final val Disabled = sbt.librarymanagement.Disabled` but it was just a companion object
- (only available in 1.3.0-x) sbt/librarymanagement#280 actually made it `final val Disabled = sbt.librarymanagement.Disabled()`, but this broke Cat's build that was calling `CrossVersion.Disabled()`
- sbt/librarymanagement#290 changed to `final def Disabled = sbt.librarymanagement.Disabled` and `object Disabled extends sbt.librarymanagement.Disabled`, but this caused #4975
- sbt/librarymanagement#316 changes back to `final val Disabled = sbt.librarymanagement.Disabled` (but because we changed the companion object in sbt/librarymanagement#290 that's ok)

**Comment by geraldstanje** [comment]
hi,

@eed3si9n i get a similar error with sbt.version = 1.2.0. i cannot upgrade to sbt 1.3.0. how to solve the error?

```
override def projectSettings: Seq[Setting[_]] = Seq(
    organization := "com.bla",
    scalaVersion := "2.12.2",
    javacOptions ++= Seq("-source", "1.7", "-target", "1.7"),
    resolvers ++= Seq[Resolver](
      ...,
    publishMavenStyle := false,
    crossPaths := false,
    crossVersion := CrossVersion.Disabled,
```

and get:
```
/Users/geri/Work/blabla/src/main/scala/com/bla/intell/sbt/IntelliBaseSettingsPlugin.scala:32:34: type mismatch;
[error]  found   : sbt.librarymanagement.Disabled.type
[error]  required: sbt.librarymanagement.CrossVersion
[error]     crossVersion := CrossVersion.Disabled,
[error]                                  ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

and this error:
```
build.sbt:5: error: type mismatch;
 found   : sbt.librarymanagement.Disabled.type
 required: sbt.librarymanagement.CrossVersion
crossVersion := CrossVersion.Disabled
                             ^
[error] Type error in expression
```

**Comment by eatkins** [comment]
> i cannot upgrade to sbt 1.3.0

Just out of curiosity, why not? If you can't upgrade to 1.3.0 because it breaks your build, is there an open issue for what is broken?

**Comment by geraldstanje** [comment]
@eatkins check: https://bintray.com/ohnosequences/sbt-plugins/sbt-s3-resolver/0.17.0?versionPath=%2Fohnosequences%2Fsbt-plugins%2Fsbt-s3-resolver%2F0.17.0

**Comment by eatkins** [comment]
Why not just upgrade sbt and disable coursier? Also, there have been many coursier bug fixes pushed since 1.3.0 so it's quite possible that the issue has been resolved in 1.3.7. The reason I'm pushing you towards 1.3.7 is because we are unlikely to make any additional releases in the 1.2.x series unless there is a really good reason.

**Comment by geraldstanje** [comment]
@eatkins i tried to build https://github.com/ohnosequences/sbt-s3-resolver with sbt 1.3.7:
- i changed line https://github.com/ohnosequences/sbt-s3-resolver/blob/master/build.sbt#L10 to 1.3.7
- i changed line https://github.com/ohnosequences/sbt-s3-resolver/blob/master/project/build.properties#L1 to 1.3.7

after i run: ```sbt clean publishLocal``` i see the following:
```
[warn] 	Note: Some unresolved dependencies have extra attributes.  Check that these dependencies exist with the requested attributes.
[warn] 		ohnosequences:nice-sbt-settings:0.9.0 (sbtVersion=1.0, scalaVersion=2.12)
```

what is Coursier?
